h/static/scripts/annotator/guest.coffee

```diff
@@ -211,21 +161,131 @@ module.exports = class Guest extends Annotator
+    self = this
+    root = @element[0]
+
+    anchors = []
+    anchoredTargets = []
+    deadHighlights = []
+
+    annotation.target ?= []
+
+    locate = (target) ->
+      options = {
+        cache: self.anchoringCache
+        ignoreSelector: '[class^="annotator-"]'
+      }
+      return self.anchoring.anchor(root, target.selector, options)
+      .then((range) -> {annotation, target, range})
+      .catch(-> {annotation, target})
+
+    highlight = (anchor) ->
+      return anchor unless anchor.range?
+      return animationPromise ->
+        range = Annotator.Range.sniff(anchor.range)
+        normedRange = range.normalize(root)
+
+        highlights = highlighter.highlightRange(normedRange)
+        rect = highlighter.getBoundingClientRect(highlights)
+
+        $(highlights).data('annotation', anchor.annotation)
+
+        anchor.highlights = highlights
+        anchor.pos =
+          left: rect.left + window.scrollX
+          top: rect.top + window.scrollY
+
+        return anchor
+
+    sync = (anchors) ->
+      annotation.$anchors = ({pos} for {pos} in anchors)
+      annotation.$orphan = anchors.length > 0
+      for anchor in anchors
+        if anchor.range?
+          annotation.$orphan = false
+
+      self.anchors = self.anchors.concat(anchors)
+      self.plugins.BucketBar?.update()
+      self.plugins.CrossFrame?.sync([annotation])
+
+    for anchor in self.anchors.splice(0, self.anchors.length)
+      if anchor.annotation is annotation
+        if anchor.range? and anchor.target in annotation.target
+          anchors.push(anchor)
+          anchoredTargets.push(anchor.target)
+        else if anchor.highlights?
+          deadHighlights.push(anchor.highlights)
+          delete anchor.highlights
+          delete anchor.range
+      else
+        self.anchors.push(anchor)
+
+    deadHighlights = Array::concat(deadHighlights...)
+    raf -> highlighter.removeHighlights(deadHighlights)
+
+    for target in annotation.target when target not in anchoredTargets
+      anchor = locate(target).then(highlight)
+      anchors.push(anchor)
+
+    Promise.all(anchors).then(sync)
+
+    return annotation
+
+  createAnnotation: (annotation = {}) ->
+    self = this
+    root = @element[0]
+
+    ranges = @selectedRanges ? []
+    @selectedRanges = null
+
+    getSelectors = (range) ->
+      options = {
+        cache: self.anchoringCache
+        ignoreSelector: '[class^="annotator-"]'
+      }
+      return self.anchoring.describe(root, range, options)
+
+    setDocumentInfo = ({metadata, uri}) ->
+      annotation.uri = uri
+      if metadata?
+        annotation.document = metadata
+
+    setTargets = ([info, selectors]) ->
+      source = info.uri
+      annotation.target = ({source, selector} for selector in selectors)
+
+    info = this.getDocumentInfo().then(setDocumentInfo)
+    selectors = Promise.all(ranges.map(getSelectors))
+    targets = Promise.all([info, selectors]).then(setTargets)
```

Comment:
Does this work? `info` is the return value of a `.then` which doesn't return anything... or rather returns two different things depending on whether there's any metadata. Either an explicit return in `setDocumentInfo` or simply doing the `.then(setDocumentInfo)` would make this clearer.